tealer/analyses/dataflow/stack_value.py

```diff
@@ -0,0 +1,229 @@
+        if len(values) > MAX_STACK_DEPTH:
+            values = values[:-MAX_STACK_DEPTH]
+
+        while values and values[0] == StackValue(TOP()):
+            values.pop()
+
+        self.values = values
+
+    def union(self, stack: "Stack") -> "Stack":
+
+        v1 = self.values
+        v2 = stack.values
+
+        min_length = min(len(v1), len(v2))
+        v1 = v1[:-min_length]
+        v2 = v2[:-min_length]
+
+        v3 = []
+        for i in range(min_length):
+            v3.append(v1[i].union(v2[i]))
+
+        return Stack(v3)
+
+    def pop(self, number: int) -> Tuple["Stack", List[StackValue]]:
+        if number == 0:
+            return Stack(self.values), []
+        if len(self.values) < number:
+            diff = number - len(self.values)
+            poped_values = list(self.values)
+            return Stack([]), [StackValue(TOP()) for _ in range(diff)] + poped_values
+
+        poped_values = self.values[:-number]
+        return Stack(self.values[: len(self.values) - number]), poped_values
+
+    def push(self, pushed_values: List[StackValue]) -> "Stack":
+        return Stack(self.values + pushed_values)
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, Stack):
+            return self.values == other.values
+        return False
+
+    def __str__(self) -> str:
+        return str([str(x) for x in self.values])
+
+
+# pylint: disable=unused-argument
+def handle_int(ins: Instruction, stack: Stack) -> List[StackValue]:
+    assert isinstance(ins, instructions.Int)
+    return [StackValue({ins.value})]
+
+
+# pylint: disable=unused-argument
+def handle_pushint(ins: Instruction, stack: Stack) -> List[StackValue]:
+    assert isinstance(ins, instructions.PushInt)
+    return [StackValue({ins.value})]
+
+
+# pylint: disable=unused-argument
+def handle_txn(ins: Instruction, stack: Stack) -> List[StackValue]:
+    assert isinstance(ins, instructions.Txn)
+    return [StackValue({ins.field})]
+
+
+# pylint: disable=unused-argument
+def handle_global(ins: Instruction, stack: Stack) -> List[StackValue]:
+    assert isinstance(ins, instructions.Global)
+    return [StackValue({ins.field})]
+
+
+special_handling: Dict[Type[Instruction], Callable[[Instruction, Stack], List[StackValue]]] = {
+    instructions.Int: handle_int,
+    instructions.PushInt: handle_pushint,
+    instructions.Txn: handle_txn,
+    instructions.Global: handle_global,
+}
+
+
+class StackValueAnalysis(AbstractDataflow[Stack]):
+    def __init__(self, teal: "Teal") -> None:
+        super().__init__(teal)
+        self.bb_in: Dict[BasicBlock, Stack] = defaultdict(lambda: Stack([StackValue(BOTTOM())]))
+        self.bb_out: Dict[BasicBlock, Stack] = defaultdict(lambda: Stack([]))
+
+        self.ins_in: Dict[Instruction, Stack] = defaultdict(lambda: Stack([]))
+        self.ins_out: Dict[Instruction, Stack] = defaultdict(lambda: Stack([]))
+
+    def _merge_predecessor(self, bb: BasicBlock) -> Stack:
+        s = Stack([])
```

Comment:
Stack union depends on the length of inputs.
```py
    def union(self, stack: "Stack") -> "Stack":

        v1 = self.values
        v2 = stack.values

        min_length = min(len(v1), len(v2))
        v1 = v1[:-min_length]
        v2 = v2[:-min_length]

        v3 = []
        for i in range(min_length):
            v3.append(v1[i].union(v2[i]))

        return Stack(v3)
```

Because initial `s = Stack([])` has length `0`, output of `union` operation will be empty stack. `_merge_predecessor` will always result in empty stack.
Correct approach would be:
```py
if not bb.prev: # zero predecessors
    return Stack([])

s = self.bb_out[bb.prev[0]]
for bb_prev in bb.prev[1:]:
    s = s.union(self.bb_out[bb_prev])

return s
```